src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+        ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+        return r;
+      }
+    } else {
+      oid_in_cache = source->get_bucket()->get_marker() + "_" + oid + "_" + std::to_string(adjusted_start_ofs) + "_" + std::to_string(obj_max_req_size);
+      //for ranged requests, for last part, the whole part might exist in the cache
+       ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << 
+      " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
+
+      if ((part_len != obj_max_req_size) && source->driver->get_policy_driver()->cachePolicy->exist_key(oid_in_cache)) {
+        // Read From Cache
+        auto completed = source->driver->get_cache_driver()->get_async(dpp, y, aio.get(), oid_in_cache, read_ofs, len_to_read, cost, id);  
+
+        source->driver->get_policy_driver()->cachePolicy->update(dpp, oid_in_cache, adjusted_start_ofs, obj_max_req_size, source->driver->get_cache_driver());
+
+        ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Info: flushing data for oid: " << oid_in_cache << dendl;
+
+        auto r = flush(dpp, std::move(completed));
+
+        if (r < 0) {
+          drain(dpp);
+          ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+          return r;
+        }
+
+      } else {
+        ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+
+        auto r = drain(dpp);
+
+        if (r < 0) {
+          ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Error: failed to drain, r= " << r << dendl;
+          return r;
+        }
+
+        break;
+      }
+    }
+
+    if (start_part_num == (num_parts - 1)) {
+      ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+      return drain(dpp);
+    } else {
+      adjusted_start_ofs += obj_max_req_size;
+    }
+
+    start_part_num += 1;
+    len -= obj_max_req_size;
+  } while (start_part_num < num_parts);
+
+
+
+  ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Fetching object from backend store" << dendl;
+
+  Attrs obj_attrs;
+  if (source->has_attrs()) {
+    obj_attrs = source->get_attrs();
+  }
+
+  if (source->is_compressed() || obj_attrs.find(RGW_ATTR_CRYPT_MODE) != obj_attrs.end() || !y) {
+    ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Skipping writing to cache" << dendl;
+    this->cb->bypass_cache_write();
+  }
+
+  if (start_part_num == 0) {
+    this->cb->set_ofs(ofs);
+  } else {
+    this->cb->set_ofs(adjusted_start_ofs);
+    ofs = adjusted_start_ofs; // redundant? -Sam
+  }
+
+  this->cb->set_ofs(ofs);
```

Comment:
If the condition on line 638 is met, then `this->cb->set_ofs(ofs)` is called twice. Is this intended?

Reply:
Well, this piece of code can definitely be reorganised, will do so, thanks